### ISSUE TITLE
New Rule: disallow parameter destructuring in Node-style callbacks

### ISSUE THREAD

**Issue by JonKrone** [eslint/eslint#10491]
**Please describe what the rule should do:**
When creating a Node-style callback function, disallow parameter destructuring of parameters after the 'error' parameter.

Destructuring in a function signature requires that an argument exists when the function is invoked or it will cause a `TypeError: Cannot destructure property ...`. Node-style callbacks are prime for this mistake because when something goes wrong, they are typically invoked with only the `Error`.

This can be a surprising code syntax bug that makes itself known during runtime failures (frequently undertested) and that a lint rule can detect. It can sneak through code review and manual review because the programmer/reviewer will check that the callback 'error' is handled and not realize the latent bug.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
const fs = require('fs')

fs.stat('invalid path', (err, { size }) => {
  if (err) {
    console.log("Looks like file didn't exist, creating it...")
    // smoothly handle error
  }

  // didn't fail, do regular things
})
```

The above code works properly when the path is correct but causes a `TypeError: Cannot destructure property 'size' of 'undefined' or 'null'.` when it is not.

**Why should this rule be included in ESLint (instead of a plugin)?**

I don't know on what grounds you guys usually classify a core/plugin rules. Where do you think this belongs?

**Comment by ljharb** [comment]
This seems like it'd be a problem with any argument that could be absent - I think the solution is to `{ size } = {}`, and I think a linting rule just to apply to node-style callbacks (a vanishing pattern in the days of promises anyways) might be a bit extreme.

**Comment by platinumazure** [comment]
I don't think this would be a good fit for core (Node-style callbacks are both hard to detect and a bit library-specific).

I think this could be great for a plugin like [`eslint-plugin-node`](https://github.com/mysticatea/eslint-plugin-node).

**Comment by JonKrone** [comment]
@ljharb I agree this is a general issue and less useful as promises become more prevalent, especially because this requires a mix of old async patterns + modern language feature.

@platinumazure Thanks for suggesting `eslint-plugin-node`, this definitely would be more at home there. I'll raise this with them!

**Comment by platinumazure** [comment]
@JonKrone Given that the proposal has been accepted in eslint-plugin-node, is it okay if we close this issue?

**Comment by platinumazure** [comment]
Closing as a new issue has been created in eslint-plugin-node.
